core_impl/src/main/java/io/opencensus/impl/tags/SerializationUtils.java

```diff
@@ -0,0 +1,195 @@
+ *                     <li>{@code <tag_key>} == tag_key_len bytes comprising tag key name
+ *                     <li>{@code <int_tag_value>} == 8 bytes, little-endian integer
+ *                   </ul>
+ *               <li>{@code (tag_field_id == 2 || tag_field_id == 3) == <tag_key_len><tag_key>}
+ *                   <ul>
+ *                     <li>{@code <tag_key_len>} == varint encoded integer
+ *                     <li>{@code <tag_key>} == tag_key_len bytes comprising tag key name
+ *                   </ul>
+ *             </ul>
+ *       </ul>
+ * </ul>
+ */
+final class SerializationUtils {
+  private SerializationUtils() {}
+
+  // TODO(songya): Currently we only support encoding on string type.
+  @VisibleForTesting static final int VERSION_ID = 0;
+  @VisibleForTesting static final int VALUE_TYPE_STRING = 0;
+  @VisibleForTesting static final int VALUE_TYPE_INTEGER = 1;
+  @VisibleForTesting static final int VALUE_TYPE_TRUE = 2;
+  @VisibleForTesting static final int VALUE_TYPE_FALSE = 3;
+
+  // Serializes a TagContext to the on-the-wire format.
+  // Encoded tags are of the form: <version_id><encoded_tags>
+  static void serialize(TagContext tags, OutputStream output) throws IOException {
+    final ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+    byteArrayOutputStream.write(VERSION_ID);
+
+    // TODO(songya): add support for value types integer and boolean
+    for (Iterator<Tag> i = tags.unsafeGetIterator(); i.hasNext(); ) {
+      Tag tag = i.next();
+
+      // TODO(sebright): Is there a better way to handle checked exceptions in function objects?
+      IOException ex =
+          tag.match(
+              new Function<TagString, IOException>() {
+                @Override
+                public IOException apply(TagString arg) {
+                  try {
+                    encodeStringTag(arg, byteArrayOutputStream);
+                    return null;
+                  } catch (IOException e) {
+                    return e;
+                  }
+                }
+              },
+              new Function<TagLong, IOException>() {
+                @Override
+                public IOException apply(TagLong arg) {
+                  throw new IllegalArgumentException("long tags are not supported.");
+                }
+              },
+              new Function<TagBoolean, IOException>() {
+                @Override
+                public IOException apply(TagBoolean arg) {
+                  throw new IllegalArgumentException("boolean tags are not supported.");
+                }
+              },
+              Functions.<IOException>throwAssertionError());
+      if (ex != null) {
+        throw ex;
+      }
+    }
+    byteArrayOutputStream.writeTo(output);
+  }
+
+  // Deserializes input to TagContext based on the binary format standard.
+  // The encoded tags are of the form: <version_id><encoded_tags>
+  static TagContextImpl deserialize(InputStream input) throws IOException {
```

Comment:
I know that this is internal but I would name this deserializeBinary or binaryDecode? Probably the reason why you have this class is because you want to share some of the code with the HTTP format?